pkg/network/http/monitor.go

```diff
@@ -171,6 +183,18 @@ func (m *Monitor) GetHTTPStats() map[Key]RequestStats {
 	return <-reply
 }
 
+func (m *Monitor) GetStats() map[string]int64 {
+	telemPtr := atomic.LoadPointer((*unsafe.Pointer)(unsafe.Pointer(&m.telemetrySnapshot)))
```

Comment:
This just copies the pointer over; what we want here to have a local copy of the actual data. So something like 
```go
dropped := atomic.LoadInt64(&m.telemetrySnapshot.dropped)
```

Reply:
Disclaimer: I am a golang beginner.

I think the proposed solution here contains a data race. Indeed, if we suppose there is a null check just before the proposed solution, nothing could theoretically prevent the pointer to be null in between the two lines, I think.

For the record, I was surprised by your example, as in a C world taking the address of an inner member would not work. It seems that [golang keeps objects alive](https://utcc.utoronto.ca/~cks/space/blog/programming/GoInteriorPointerGC) even though we're referencing inner pointers. However, issue referenced before still remains.

Besides, even if that proposition would work, I think it doesn't evolve well. If we would have another field to be transferred for telemetry, then we would basically need two of those lines. However, nothing prevents the pointer to change in between either, which means we'd have inconsistent data.

As per [this](https://groups.google.com/g/golang-nuts/c/yNis7bQG_rY/m/yaJFoSx1hgIJ) (if that's still working like that), it seems that unsafe pointers referencing golang objects are correctly handled by GC, which means that even though the pointer would change right after this line, it would still keep the telemetry object alive.

I might misunderstand some golang internals though, but I think we should stick to how it currently is.

Reply:
As an overall goal, we should avoid `unsafe` where possible.

> Indeed, if we suppose there is a null check just before the proposed solution, nothing could theoretically prevent the pointer to be null in between the two lines

If you initialize `m.telemetrySnapshot` when `Monitor` is created, there is no danger of a null pointer. You also don't have to worry about GC.

> I think it doesn't evolve well. If we would have another field to be transferred for telemetry, then we would basically need two of those lines

I think that is OK actually. If you want to go down the full struct swap route, I would look into a `sync/atomic.Value` rather than the `unsafe.Pointer` stuff.

Reply:
> As an overall goal, we should avoid `unsafe` where possible.

I agree with that. However, here we're only type casting to make it compliant with a package interface we're using. It feels fairly ok in that case IMO. However, `atomic.Value` could indeed fix this if you feel it is really an issue using `unsafe` only for doing casts in a function call.

> If you initialize `m.telemetrySnapshot` when `Monitor` is created, there is no danger of a null pointer. You also don't have to worry about GC.

The GC part was purely informative here. `m.telemetrySnapshot` **is not** initialised when `Monitor` is created, so it doesn't stand anyways. I think it would not make sense to initialise it to anything at the moment, as what we want to make sure is that the data we're fetching in `getStats` has been previously synchronously fetched with `getHTTPStats`.

I think my first example to why there is a data race in the proposed solution might have not been clear enough (although first mentioned example **still stands**).
Consider [this](https://go.dev/play/p/1Y9-a8ujTGn), which reproduces more simply our exact case here. Compile it with `go build -race main.go`, and you'll see that it triggers a data race.
As stated in my previous comment, it is because you cannot store a pointer atomically on one hand, and fetch it non-atomically on the other hand. This is not semantically correct in a pure concurrent programming point of view, and on some architecture, you could even end up with a segmentation fault (most likely not on x86_64 and arm64 tbh).

The problem is, if we want to go down the path of fetching the values separately, my other concern still stands. Even with code that doesn't contain data race, we could end up with inconsistent data if we were to have a second telemetry field to check. They could be coming from two different fetches.

Reply:
To my point about `unsafe`, this is actually incorrect code: `(*unsafe.Pointer)(unsafe.Pointer(&test))`. You are casting an `unsafe.Pointer` to a `*unsafe.Pointer`, which it is not. I think the atomic API wants you to have a stable `unsafe.Pointer` that you then do `&` on. See the tests: https://github.com/golang/go/blob/7900576bac4630bbeec7f4f1aa4b1cb0d51bd8a1/src/sync/atomic/atomic_test.go#L605-L625

It sounds like we either want to use an `atomic.Value` or a mutex would also accomplish what we want.

Reply:
Right, although according to [this](https://go101.org/article/unsafe.html) (fact 5), and [this](https://github.com/golang/go/blob/7900576bac4630bbeec7f4f1aa4b1cb0d51bd8a1/src/sync/poolqueue.go#L225), it seems that `*unsafe.Pointer` and `unsafe.Pointer` are safely convertible to each other, as per the type definition of `unsafe.Pointer`, it really represents anything, so one can see it as a C `void *`. We don't really care what's the actual representation of it. It seems that `checkptr` doesn't see any issue with it either (`-gcflags=all=-d=checkptr`).

Do we agree that doing it on a per-field basis is not the way to go ? I think using a mutex here is a bit over-engineered, so if we really don't want to go down the `unsafe.Pointer` path, I guess `atomic.Value` is the way to go...

Reply:
I understand the concern about `unsafe.Pointer`. I am just wondering whether it makes sense to introduce the mutex in the go-routine that's also in charge of extracting http data. I'm mostly comparing the non-blocking vs the possibly blocking solution here, especially for the use case we have here, and given we'll end up with the exact same functional result.
Maybe then the solution is to move towards `atomic.Value` ? It seems to be roughly doing what we're doing here with the pointer, but it would avoid us using unsafe pointers.

Reply:
If you use a mutex, I would constrain the scope of the mutex to as small as possible. I think adding a new function that reads the `delta` variable would be sufficient, along with using the mutex in `GetStats`. The contention on the mutex will be minimal, especially considering these functions are called once every 30s.

We could use `atomic.Value` if you like. Looking at how it is implemented, it feels a bit overkill, but I don't have strong feelings one way or the other.

I think a lot of your concerns would be worth more examination if this was a hot path, or something with a lot of contention or concurrency. This is not that 😄 

We should make sure it is correct and safe, but optimization is not a concern here.